Python/crossinterp.c

```diff
@@ -713,6 +713,76 @@ _bool_shared(PyThreadState *tstate, PyObject *obj,
     return 0;
 }
 
+struct _shared_tuple_data {
+    Py_ssize_t len;
+    _PyCrossInterpreterData ** data;
+};
+
+static PyObject *
+_new_tuple_object(_PyCrossInterpreterData *data)
+{
+    struct _shared_tuple_data *shared = (struct _shared_tuple_data *)(data->data);
+    PyObject *tuple = PyTuple_New(shared->len);
+    if (tuple == NULL) {
+        return NULL;
+    }
+
+    for (Py_ssize_t i = 0; i < shared->len; i++) {
+        PyObject *item = _PyCrossInterpreterData_NewObject(shared->data[i]);
+        if (item == NULL){
+            Py_DECREF(tuple);
+            return NULL;
+        }
+        PyTuple_SET_ITEM(tuple, i, item);
+    }
+    return tuple;
+}
+
+static void
+_tuple_shared_free(void* data)
+{
+    struct _shared_tuple_data *shared = (struct _shared_tuple_data *)(data);
+    for (Py_ssize_t i = 0; i < shared->len; i++) {
+        _PyCrossInterpreterData_Release(shared->data[i]);
+    }
```

Comment:
Make sure we skip entries that aren't set, just like `tupledealloc()` does:

```suggestion
    struct _shared_tuple_data *shared = (struct _shared_tuple_data *)(data);
    for (Py_ssize_t i = 0; i < shared->len; i++) {
        if (shared->data[i] != NULL) {
            _PyCrossInterpreterData_Release(shared->data[i]);
        }
    }
```

or, if we are confident that a `NULL` entry indicates all of the rest will be `NULL`:

```suggestion
    struct _shared_tuple_data *shared = (struct _shared_tuple_data *)(data);
    for (Py_ssize_t i = 0; i < shared->len; i++) {
        if (shared->data[i] != NULL) {
            break;
        }
    }
```

However, I'd probably go with the first one, just to play it safe.

Reply:
We also need to delete the items after they are released:

```suggestion
    struct _shared_tuple_data *shared = (struct _shared_tuple_data *)(data);
    for (Py_ssize_t i = 0; i < shared->len; i++) {
        if (shared->data[i] != NULL) {
            _PyCrossInterpreterData_ReleaseAndRawFree(shared->data[i]);
            shared->data[i] = NULL;
        }
    }
```

or, since we always expect all items to come from the same interpreter as the tuple:

```suggestion
    struct _shared_tuple_data *shared = (struct _shared_tuple_data *)(data);
#ifndef NDEBUG
    int64_t interpid = PyInterpreterState_GetID(_PyInterpreterState_GET()));
#endif
    for (Py_ssize_t i = 0; i < shared->len; i++) {
        if (shared->data[i] != NULL) {
            assert(shared->data[i]->interpid == interpid);
            _PyCrossInterpreterData_Release(shared->data[i]);
            PyMem_RawFree(shared->data[i]);
            shared->data[i] = NULL;
        }
    }
```